src/models/Uploads.php

```diff
@@ -251,27 +250,44 @@ public function getStorageFromLongname(string $longname): int
         return (int) $req->fetchColumn();
     }
 
-    /**
-     * This function will not remove the files but set them to "deleted" state
-     * A manual purge must be made by sysadmin if they wish to really remove them.
-     */
-    private function nuke(): bool
+    public function getIdFromLongname(string $longname): int
     {
-        return $this->update(new UploadParams((string) self::STATE_DELETED, 'state'));
+        $sql = 'SELECT id FROM uploads WHERE long_name = :long_name LIMIT 1';
+        $req = $this->Db->prepare($sql);
+        $req->bindParam(':long_name', $longname, PDO::PARAM_STR);
+        $this->Db->execute($req);
+        return (int) $req->fetchColumn();
     }
 
     /**
      * Attached files are immutable (change history is kept), so the current
      * file gets its state changed to "archived" and a new one is added
      */
-    private function replace(UploadedFile $file): bool
+    public function replace(UploadParamsInterface $params): array
     {
+        $this->Entity->canOrExplode('write');
         // read the current one to get the comment
         $upload = $this->read(new ContentParams());
-        $params = new CreateUpload($file->getClientOriginalName(), $file->getPathname(), $upload['comment']);
-        $this->create($params);
+        $replaceRes = $this->update(new UploadParams((string) self::STATE_ARCHIVED, 'state'));
+        
+        if (!$replaceRes) {
+            throw new ImproperActionException(_('Error while replacing file!'));
```

Comment:
I don't think this needs checking. MySQL will throw an exception if there is an issue during update. Also it avoids adding another string to translate.

Reply:
> MySQL will throw an exception

I don't know, ideally the two SQL statements (update and insert) would be wrapped in a transaction.

> it avoids adding another string to translate

Makes sense